### PR TITLE
Add: First shot at content link url

### PR DESCRIPTION
**Pull request by fsbraun**
This pull requests adds a field ``content_link_url`` to the Post model to link an external source to the blog post, e.g. a video or audio resource. If the blog post does not have a main image the ``content_link_url`` can provide an alternative, e.g. a still of the video. 

Template designers can use the template link to incorporate video or audio players, light boxes etc. to easily access the linked content.

This feature should make video blogs or podcasts easily managed by djangocms blog.

### CODE REVIEW

File: docs/features.rst [hunk 1]
```diff
@@ -265,3 +266,20 @@ The shariff files ``js/shariff.min.js`` and ``css/shariff.min.css`` will need to
...
+Linked content
+**************
+
+Each blog entry can have an image assigned for the blog post. The image is typically presented in the post details and a thumbnail in the post list.
+
+Dangocms-blog, however, also allows to link external content to a blog entry. This is done by enternig a valid URL in the field ``content link`` when editing a blog entry.
+
+Such a content link could be a featured video, a podcast episode or just a link to an image, or any content you deem important for the blog entry.
+
+If no main image is assigned to the blog, djangocms will try to see if it can gereate an image and/or thumbnail from the link. To this end it uses regular expressions that will for example turn a YouTube video url into an url for a still or thumbnail. An audiofile url might be turned into a link for album art or image of the podcast narrator. Default settings will only return stills for YouTube videos and jpeg, png and tiff images. See :ref:`settings` for more information on how to configure this feature.
```
Comment (yakky) [inline docs/features.rst]: `s/gereate/generate/`

File: docs/settings.rst [hunk 2]
```diff
@@ -99,6 +99,19 @@ Global Settings
...
 * BLOG_POST_TEXT_CKEDITOR: Configuration for the CKEditor of the post content field
+* BLOG_LINKED_IMAGE and BLOG_LINKED_THUMBNAIL: Both provide a list of 2-tuples djangocms-blog will use to create a valid url from ``linked_content_url`` field. The format is ``('regular expression'), 'url')``. Assume, e.g., that ``linked_content_url`` contains a link to a YouTube video. Then the entry 
+
+		('^https://www.youtube.com/watch\?v=(?P<vid>[-\w]+)$',
+		'https://img.youtube.com/vi/%(vid)s/hqdefault.jpg')
+              
+will create a thumbnail link by matching and finding the video id ``vid``and incorporating it in an appropriate link. In addition to parsed parameters (``vid`` in the YouTube example) the full original url will be passed to the pattern in the second part of the 2-tuple. If the url is callable, e.g. a lambda expression, djangocms-blog will call the function and pass the parameters and the original url. This would allow to call API if neccessary. This illustrative example would call the Vimeo API to get a large thumbnail for a video:
```
Comment (yakky) [inline docs/settings.rst]: `s/neccessary/necessary/`

File: docs/features.rst [hunk 3]
```diff
@@ -218,6 +218,7 @@ To add the blog Sitemap, add the following code to the project ``urls.py``::
         }),
     )
 
+.. _social_sahres:
```
Comment (yakky) [inline docs/features.rst]: `s/social_sahres/social_shares/`

File: djangocms_blog/settings.py [hunk 4]
```diff
@@ -147,5 +147,25 @@ def get_setting(name):
         'BLOG_PLUGIN_TEMPLATE_FOLDERS': getattr(
             settings, 'BLOG_PLUGIN_TEMPLATE_FOLDERS', (('plugins', _('Default template')),)),
 
+        'BLOG_LINKED_IMAGE': (
```
Comment (yakky) [inline djangocms_blog/settings.py]: Do we really need this?
wouldn't be easier for the user to upload the cover image by using the existing fields?
I suspect the rules below only applies to youtube and they are not generic enough to justify the added logic

**Comment by fsbraun** [comment]
I'm open for discussion. I see the user entering a url for linked content and the blog app generating thumbnails and images automatically. Anything else would put the burden of getting and uploading the images/thumbnails on the editor. 

I see two paths: 
(a) Introduce two fields (say, ``linked_content_image_url`` and ``linked_content_thumbnail_url`` in the model. These should be hidden and automatically generated up change of the ``linked_content_url``.  A separate module could handle required API calls to content services like YouTube, Vimeo, and what have you. The API module could be extended over time. The regex blog config entries will disappear.

(b) Keep the model thin and generate the image/thumbnail urls on the fly. The regex-based solution does exactly this. The underlying assumption is that there is a relatively simple rule on how to obtain the image/thumbnail urls (like with YouTube if you know the video id). You can do API calls (as the Vimeo example shows, where video id and thumbnail id differ) but they will generate far too much latency in a real world. On the plus side is that there is no need for a module handling API calls keeping the logic relatively thin (maybe not too readable, though, if you are not into regex).

I am happy to go down route (a) because it is more versatile. However, we should think on what content services to include. I see the need for coverage of video and audio besides images (which are nicely served by djangocms-blog). 

My use case is **YouTube**, **Vimeo** would be an obvious alternative. **SoundCloud** might be relevant for podcasters. **In your opinion, what other content services might be important?** 

@yakky What are your thoughts on this?

**Comment by yakky** [comment]
@fsbraun I am not overlooking this, just pondering the right balance between the different options here. I will get back to you soon
@nephila/maintainers any thoughts on this?

**Comment by fsbraun** [comment]
@yakky @nephila/maintainers  To test the concept I added a new file ``linked_content.py`` that creates the urls for the image and thumbnail a given content link automatically. It recognizes jpeg, tiff and png images, YouTube videos, Vimeo videos and Soundcloud tracks (for the latter an API key is required to be set in ``SOUNDCLOUD_APIKEY``). 

With the adjusted templates a post image always overrides the linked content. Also, it only displays the image/thumbnail. A real-world use case would also allow interaction like playing a video or sound bite.

The unhandy regex settings are gone.

**Comment by yakky** [comment]
@fsbraun do you mind if I will propose a different approach in a new PR?
I would like something more "pluggable" and configurable by the end sure, while limiting the code added directly to the main `Post` class

**Comment by fsbraun** [comment]
@yakky I would appreciate your foresight with a more scalable structure. Having ``LinkedContent`` as a mixin was targeted to reduce the code added to the ``Post`` class. 

This being said: The more versatile the code the better. I will probably need the feature within the next 6 months or so. Great if we manage, otherwise I will fork for the meantime. Let me know how I can support.

**Comment by yakky** [comment]
@fsbraun sorry it took me so long to work on this, but I elaborated my thoughs in #490 It seems to address your needs, and still keep the djangocms-blog code lean and open to customization

**Pull request closed without merging**